pkg/dcgmexporter/expcollector.go

```diff
@@ -0,0 +1,249 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.  All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package dcgmexporter
+
+import (
+	"fmt"
+	"io"
+	"maps"
+	"slices"
+	"sync"
+	"sync/atomic"
+	"text/template"
+	"time"
+
+	"github.com/NVIDIA/go-dcgm/pkg/dcgm"
+	"github.com/sirupsen/logrus"
+)
+
+var expMetricsFormat = `
+
+{{- range $counter, $metrics := . -}}
+# HELP {{ $counter.FieldName }} {{ $counter.Help }}
+# TYPE {{ $counter.FieldName }} {{ $counter.PromType }}
+{{- range $metric := $metrics }}
+{{ $counter.FieldName }}{gpu="{{ $metric.GPU }}",{{ $metric.UUID }}="{{ $metric.GPUUUID }}",device="{{ $metric.GPUDevice }}",modelName="{{ $metric.GPUModelName }}"{{if $metric.MigProfile}},GPU_I_PROFILE="{{ $metric.MigProfile }}",GPU_I_ID="{{ $metric.GPUInstanceID }}"{{end}}{{if $metric.Hostname }},Hostname="{{ $metric.Hostname }}"{{end}}
+
+{{- range $k, $v := $metric.Labels -}}
+	,{{ $k }}="{{ $v }}"
+{{- end -}}
+{{- range $k, $v := $metric.Attributes -}}
+	,{{ $k }}="{{ $v }}"
+{{- end -}}
+
+} {{ $metric.Value -}}
+{{- end }}
+{{ end }}`
+
+// Collector interface
+type Collector interface {
+	GetMetrics() (MetricsByCounter, error)
+	Cleanup()
+}
+
+var getExpMetricTemplate = sync.OnceValue(func() *template.Template {
+	return template.Must(template.New("expMetrics").Parse(expMetricsFormat))
+
+})
+
+func encodeExpMetrics(w io.Writer, metrics MetricsByCounter) error {
+	tmpl := getExpMetricTemplate()
+	return tmpl.Execute(w, metrics)
+}
+
+var expCollectorFieldGroupIdx atomic.Uint32
+
+type expCollector struct {
+	sysInfo             SystemInfo
+	counter             Counter
+	hostname            string
+	config              *Config
+	labelDeviceFields   []dcgm.Short
+	counterDeviceFields []dcgm.Short
+	labelsCounters      []Counter
+	cleanups            []func()
+	fieldValueParser    func(val int64) []int64
+	labelFiller         func(map[string]string, int64)
+	windowSize          int
+	transformations     []Transform
```

Comment:
Does this needs to be a slice, I see only a single podMapper return.

Reply:
It is supposed that we may have more than one transformer.